src/cloud-api-adaptor/pkg/securecomms/wnssh/wnssh.go

```diff
@@ -0,0 +1,386 @@
+	ci := &SshClientInstance{
+		sid:          sid,
+		publicKey:    serverSshPublicKeyBytes,
+		ppAddr:       ppAddr,
+		sshClient:    c,
+		ctx:          ctx,
+		cancel:       cancel,
+		inboundPorts: make(map[string]string),
+	}
+
+	for _, tag := range c.inboundStrings {
+		name, inPort, err := ci.inbounds.Add(tag, &ci.wg)
+		if err != nil {
+			log.Printf("failed to add inbound: %v", err)
+		}
+		ci.inboundPorts[name] = inPort
+	}
+	for _, tag := range c.outboundStrings {
+		ci.outbounds.Add(tag)
+	}
+
+	return ci
+}
+
+func (ci *SshClientInstance) Start() error {
+	if !ci.k8sPhase {
+		// Attestation Phase
+		log.Println("Attstation Phase: Starting")
+		if err := ci.StartAttestation(); err != nil {
+			return fmt.Errorf("attstation Phase failed: %v", err)
+		}
+		log.Println("Attstation Phase: Done")
+		ci.k8sPhase = true
+	}
+
+	// Kubernetes Phase
+	ci.wg.Add(1)
+	go func() {
+		defer ci.wg.Done()
+		restarts := 0
+		for {
+			select {
+			case <-ci.ctx.Done():
+				log.Printf("Kubernetes Phase: Done")
+				return
+			default:
+				log.Printf("Kubernetes Phase: Starting (Number of restarts %d)", restarts)
+				if err := ci.StartKubernetes(); err != nil {
+					log.Printf("Kubernetes Phase: failed: %v", err)
+				}
+				time.Sleep(time.Second)
+				restarts += 1
+			}
+		}
+	}()
+	return nil
+}
+
+func (ci *SshClientInstance) StartKubernetes() error {
+	ctx, cancel := context.WithCancel(ci.ctx)
+	defer cancel()
+	peer := ci.StartSshClient(ctx, sshproxy.KUBERNETES, ci.publicKey, ci.sid)
+	if peer == nil {
+
+		return fmt.Errorf("Kubernetes Phase: failed StartSshClient")
+	}
+
+	peer.AddOutbounds(ci.outbounds)
+	err := peer.AddInbounds(ci.inbounds)
+	if err != nil {
+		peer.Close("Inbounds failed")
+		peer = nil
+		return fmt.Errorf("inbounds failed: %w", err)
+	}
+	peer.Wait()
+	return nil
+}
+
+func (ci *SshClientInstance) StartAttestation() error {
+	ctx, cancel := context.WithCancel(ci.ctx)
+	defer cancel()
+	peer := ci.StartSshClient(ctx, sshproxy.ATTESTATION, nil, ci.sid)
+	if peer == nil {
+		return fmt.Errorf("Attestation Phase: failed StartSshClient")
+	}
+	peer.AddOutbounds(ci.outbounds)
+	err := peer.AddInbounds(ci.inbounds)
+	if err != nil {
+		peer.Close("Inbounds failed")
+		peer = nil
+		return fmt.Errorf("inbounds failed: %w", err)
+	}
+	peer.Wait()
+	if !peer.IsUpgraded() {
+		return fmt.Errorf("Attestation PHASE closed without being upgraded")
+	}
+	return nil
+}
+
+func (ci *SshClientInstance) StartSshClient(ctx context.Context, phase string, publicKey []byte, sid string) *sshproxy.SshPeer {
+	config := &ssh.ClientConfig{
+		HostKeyCallback: func(hostname string, remote net.Addr, key ssh.PublicKey) error {
+			if len(publicKey) == 0 {
+				log.Printf("%s Phase: ssh Client skip validating server's HOST KEY (%s) during attestation", phase, key.Type())
+				return nil
+			}
+			if !bytes.Equal(key.Marshal(), ci.publicKey) {
+				log.Printf("%s Phase: ssh Client HOST KEY mismatch - %s", phase, key.Type())
```

Comment:
Shouldn't that be elevated to error or warning level?

Reply:
This is not an error (not even a warning IMHO) - this can happen in normal operation when someone without proper credentials try to connect to the system. It needs to be logged as info and in most cases ignored.